pkg/kubelet/cm/cpumanager/cpu_manager_test.go

```diff
@@ -533,16 +534,157 @@ func TestCPUManagerAddWithInitContainers(t *testing.T) {
-				testCase.description, testCase.stDefaultCPUSet.Difference(cumCSet), state.defaultCPUSet)
+				testCase.description, testCase.stDefaultCPUSet.Difference(appCtrCumCSet), state.defaultCPUSet)
 		}
 	}
 }
 
+func TestCPUManagerAddWithMultiplePods(t *testing.T) {
+	testCases := []struct {
+		description      string
+		topo             *topology.CPUTopology
+		numReservedCPUs  int
+		initContainerIDs [][]string
+		containerIDs     [][]string
+		stAssignments    state.ContainerCPUAssignments
+		stDefaultCPUSet  cpuset.CPUSet
+		pod              []*v1.Pod
+		expInitCSets     [][]cpuset.CPUSet
+		expCSets         [][]cpuset.CPUSet
+	}{
+		{
+			description:     "Multi Init, Multi App Container Split CPUs",
+			topo:            topoSingleSocketHT,
+			numReservedCPUs: 0,
+			stAssignments:   state.ContainerCPUAssignments{},
+			stDefaultCPUSet: cpuset.NewCPUSet(0, 1, 2, 3, 4, 5, 6, 7),
+			initContainerIDs: [][]string{
+				{"initFakeID-1", "initFakeID-2"},
+				{"initFakeID-1"},
+			},
+			containerIDs: [][]string{
+				{"appFakeID-1", "appFakeID-2"},
+				{"appFakeID-1"},
+			},
+			pod: []*v1.Pod{
+				makeMultiContainerPod(
+					[]struct{ request, limit string }{
+						{"2000m", "2000m"},
+						{"4000m", "4000m"},
+					},
+					[]struct{ request, limit string }{
+						{"3000m", "3000m"},
+						{"2000m", "2000m"},
+					}),
+				makeMultiContainerPod(
+					[]struct{ request, limit string }{
+						{"3000m", "3000m"},
+					},
+					[]struct{ request, limit string }{
+						{"3000m", "3000m"},
+					}),
+			},
+			expInitCSets: [][]cpuset.CPUSet{
+				{
+					cpuset.NewCPUSet(0, 4),
+					cpuset.NewCPUSet(0, 4, 1, 5),
+				},
+				{
+					cpuset.NewCPUSet(3, 5, 7),
+				},
+			},
+			expCSets: [][]cpuset.CPUSet{
+				{
+					cpuset.NewCPUSet(0, 1, 4),
+					cpuset.NewCPUSet(2, 6),
+				},
+				{
+					cpuset.NewCPUSet(3, 5, 7),
+				},
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		policy, _ := NewStaticPolicy(testCase.topo, testCase.numReservedCPUs, cpuset.NewCPUSet(), topologymanager.NewFakeManager())
+
+		state := &mockState{
+			assignments:   testCase.stAssignments,
+			defaultCPUSet: testCase.stDefaultCPUSet,
+		}
+
+		mgr := &manager{
+			policy:            policy,
+			state:             state,
+			containerRuntime:  mockRuntimeService{},
+			containerMap:      containermap.NewContainerMap(),
+			podStatusProvider: mockPodStatusProvider{},
+			sourcesReady:      &sourcesReadyStub{},
+			activePods: func() []*v1.Pod {
+				return testCase.pod
+			},
+		}
+		appCtrCumCSet := cpuset.NewCPUSet()
+
+		for ix, pod := range testCase.pod {
+			containers := append(
+				pod.Spec.InitContainers,
+				pod.Spec.Containers...)
+
+			containerIDs := append(
+				testCase.initContainerIDs[ix],
+				testCase.containerIDs[ix]...)
+
+			expCSets := append(
+				testCase.expInitCSets[ix],
+				testCase.expCSets[ix]...)
+
+			for i := range containers {
+				err := mgr.Allocate(pod, &containers[i])
```

Comment:
Something I don’t quite follow regarding how this works:

Init containers are allocated CPUs by the call to mgr.Allocate(), correct?

How do the CPUs allocated to init containers get “released” so they can be allocated to main containers? I don’t see an explicit call that does that. Is there something implicit happening here? Just trying to understand better how the test works.